README.md
# MagickBytes

A tiny library for inferring MIME types from magic bytes 

## Installation

The package can be installed
by adding `magick_bytes` to your list of dependencies in `mix.exs`:

```elixir
def deps do
  [
    {:magick_bytes, "~> 0.2.0"}
  ]
end
```

## Usage

### Common MIME types

To guess the MIME type of a file, just pass its
path as an argument to the mime/1 fn. MagickBytes
will open the file, read from it and try to
guess the MIME type:

```elixir
MagickBytes.mime("/path/to/my/document.pdf")
> {:ok, "application/pdf"}
```

### Custom matchers

If the file that you are trying to match is not
covered by this library or is otherwise a custom
format, you can set up your own matcher:

```elixir
defmodule ObscureFileMatcher do
  use MagickBytes.CustomMatcher,

    # Matches a byte signature (magic bytes)
    signature: <<1, 2, 3, 4, 5, 6>>,

    # Matches an ASCII string in the file
    string: "EBML",

    # The mime string to be returned if
    # a match is found
    mime: "application/obscure_format"
end
```

You can then pass to your custom matcher
a group of bytes and it will try a match:

```elixir
ObscureFileMatcher.mime(<<1, 2, 3, 4, 5, 6, 69, 66, 77, 76>>)
> "application/obscure_format"

ObscureFileMatcher.mime(<<42, 13, 69>>)
> false
```

All conditions are checked and must pass to return a match.

## Security considerations

Remember that every magic is illusion and can be trivially spoofed. In other words, never do anything like this:

```elixir
case MagickBytes.mime(secret_file) do
  {:ok, "application/secret_file"} -> give_password()
  _ -> unauthorized()
end
```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.2.0]

### Added

 - Custom matchers (#1)

## [0.1.0]

- Initial release

test/matcher_test.exs
defmodule MagickBytes.MatcherTest do
  use ExUnit.Case, async: true
  use ExUnitProperties

  property ~s|match/1 matches or asks for more bytes, returning bytes
      that were matched on previous run| do
    check all bytes <- binary() do
      case MagickBytes.Matcher.match(bytes) do
        {:match, _something} ->
          assert true

        {:continue, :no_match, bytes_matched} ->
          assert bytes_matched == bytes

        _ ->
          raise "should be unreachable"
      end
    end
  end
end


test/magick_bytes_test.exs
defmodule MagickBytesTest do
  use ExUnit.Case, async: true
  doctest MagickBytes

  describe "mime/1" do
    test "should match zip" do
      assert {:ok, "application/zip"} == MagickBytes.mime("./test/archives/zipfile.zip")
    end

    test "should match matroska" do
      assert {:ok, "video/x-matroska"} ==
               MagickBytes.mime("./test/archives/test1.mkv")
    end

    test "should match webm" do
      assert {:ok, "video/webm"} ==
               MagickBytes.mime("./test/archives/big-buck-bunny_trailer.webm")
    end

    test ~s|should return {:error, :no_match} after reading an empty file
    and failing to find a match| do
      assert {:error, :no_match} == MagickBytes.mime("./test/archives/empty")
    end
  end
end


test/custom_matcher_test.exs
defmodule MagickBytes.CustomMatcherTest do
  use ExUnit.Case, async: true

  defmodule SimpleMatcher do
    use MagickBytes.CustomMatcher,
      signature: <<23, 23>>,
      mime: "application/custom-match"
  end

  defmodule StringMatcher do
    use MagickBytes.CustomMatcher,
      signature: <<23, 23>>,
      string: "hello",
      mime: "application/custom-match"
  end

  defmodule AnyMatcher do
    use MagickBytes.CustomMatcher, mime: "application/custom-match"
  end

  test "should match <<match>>" do
    assert SimpleMatcher.mime(<<23, 23>>) == "application/custom-match"
    refute SimpleMatcher.mime(<<23, 24>>)
  end

  test "should match <<match>> and <<string>>" do
    assert StringMatcher.mime(<<23, 23, 0, 0, 0, 104, 101, 108, 108, 111>>) ==
             "application/custom-match"

    refute StringMatcher.mime(<<23, 23, 0, 1, 0, 105, 101, 107, 108, 111>>)
  end

  test "should match anything" do
    assert AnyMatcher.mime(<<23, 23>>) == "application/custom-match"
  end
end


lib/magick_bytes.ex
defmodule MagickBytes do
  @moduledoc """
  A tiny library for inferring
  MIME types from magic bytes.
  """

  alias MagickBytes.Matcher

  @doc """
  Tries to infer the MIME type
  of the file at `path` from
  its magic bytes.

  ## Examples

      iex> MagickBytes.mime("./test/archives/my_gzip_file_disguised_as.txt")
      {:ok, "application/gzip"}

  """
  @spec mime(path :: Path.t()) :: {:ok, String.t()} | {:error, atom()}
  def mime(path) when is_binary(path) do
    with {:ok, file} <- File.open(path, [{:read_ahead, 640}]),
         result <- try_match(file),
         :ok <- File.close(file) do
      result
    end
  end

  defp try_match(file) do
    with {:ok, byte} <- :file.pread(file, 0, 1),
         {:continue, :no_match, _bytes} <- Matcher.match(byte) do
      try_match(file, 1, byte)
    else
      :eof -> {:error, :no_match}
      {:error, :einval} -> {:error, :no_match}
      {:match, match} -> {:ok, match}
    end
  end

  defp try_match(_file, 640, _bytes), do: {:error, :no_match}

  defp try_match(file, current_offset, bytes_read) do
    with {:ok, byte} <- :file.pread(file, current_offset, 1),
         {:continue, :no_match, bytes} <-
           Matcher.match(<<bytes_read::binary>> <> <<byte::binary>>) do
      try_match(file, current_offset + 1, bytes)
    else
      :eof -> {:error, :no_match}
      {:error, :einval} -> {:error, :no_match}
      {:match, match} -> {:ok, match}
      error -> error
    end
  end
end


lib/custom_matcher.ex
defmodule MagickBytes.CustomMatcher do
  defmacro __using__(opts) do
    mime = Keyword.fetch!(opts, :mime)
    match = Keyword.get(opts, :signature, "")
    bin = Keyword.get(opts, :string, nil)

    quote do
      def mime(bytes_to_match) do
        signature(bytes_to_match) && string(bytes_to_match) && unquote(mime)
      end

      defp signature(bytes_to_match) do
        case unquote(match) do
          "" -> true
          any -> match_signature(bytes_to_match)
        end
      end

      defp match_signature(<<unquote(match), _rest::binary>>), do: true
      defp match_signature(_), do: false

      defp string(bytes_to_match) do
        case unquote(bin) do
          nil ->
            true

          string ->
            pattern_match(bytes_to_match, string)
        end
      end

      defp pattern_match(binary, pattern) do
        case :binary.match(binary, pattern) do
          :nomatch -> false
          _ -> true
        end
      end
    end
  end
end


lib/matcher.ex
defmodule MagickBytes.Matcher do
  # 1F 8B
  def match(<<31, 139>>) do
    {:match, "application/gzip"}
  end

  def match(<<80, 75>>) do
    {:match, "application/zip"}
  end

  # FF D8 FF
  def match(<<255, 216, 255>>) do
    {:match, "image/jpeg"}
  end

  # ED AB EE DB
  def match(<<237, 171, 238, 219>>) do
    {:match, "application/x-rpm"}
  end

  # 4F 67 67 53
  def match(<<79, 103, 103, 83>>) do
    {:match, "application/ogg"}
  end

  # 25 50 44 46 2D
  def match(<<37, 80, 68, 70, 45>>) do
    {:match, "application/pdf"}
  end

  # 52 49 46 46
  # ?? ?? ?? ??
  # 41 56 49 20
  def match(<<82, 73, 70, 70, _::8, _::8, _::8, _::8, 65, 86, 73, 32>>) do
    {:match, "audio/x-avi"}
  end

  # 52 49 46 46
  # ?? ?? ?? ??
  # 57 41 56 45
  def match(<<82, 73, 70, 70, _::8, _::8, _::8, _::8, 87, 65, 86, 69>>) do
    {:match, "audio/x-wav"}
  end

  # 42 4d ?? ?? ?? ?? ?? ??
  # ?? ?? ?? ?? 00 00 ?? 00
  # http://fileformats.archiveteam.org/wiki/BMP#Identifiers
  def match(<<66, 77, _::8, _::8, _::8, _::8, _::8, _::8, _::8, _::8, _::8, 0, 0, _::8, 0>>) do
    {:match, "application/x-ms-bmp"}
  end

  # 53 51 4C 69 74 65 20 66
  # 6F 72 6D 61 74 20 33 00
  def match(<<83, 81, 76, 105, 116, 101, 32, 102, 111, 114, 109, 97, 116, 32, 51, 0>>) do
    {:match, "application/vnd.sqlite3"}
  end

  # 1A 45 DF A3
  # EBML
  def match(<<26, 69, 223, 163, rest::binary>> = bytes) do
    cond do
      pattern_match(rest, <<119, 101, 98, 109>>) ->
        {:match, "video/webm"}

      pattern_match(rest, <<109, 97, 116, 114, 111, 115, 107, 97>>) ->
        {:match, "video/x-matroska"}

      true ->
        {:continue, :no_match, bytes}
    end
  end

  def match(bytes) do
    {:continue, :no_match, bytes}
  end

  defp pattern_match(binary, pattern) do
    case :binary.match(binary, pattern) do
      :nomatch -> false
      _ -> true
    end
  end
end
